1.0.3

elich11 opened the pull request:
1.0.3

elich11 committed:
device rotation does not re-create connection

elich11 committed:
device rotation does not re-create connection

jo-elimu reviewed (commented):
Will this application have a visible UI at all? Or just temporarily for testing purposes? Perhaps it should have a UI (not available to the children) listing connected nearby devices?

elich11 commented:
I agree, Jo. 
Admin UI is needed, and not only to see connected devices,
At the beginning (at least) there will be issues that will need human intervention.
That said, We should design and code it to auto manage it as much as possible.

elich11 committed:
initial work,  discover works (above Android no external lib)

elich11 committed:
peer list working (no UI)

jo-elimu commented on app/src/main/AndroidManifest.xml:
```diff
@@ -9,15 +9,26 @@
-        android:label="@string/app_name" android:supportsRtl="true" android:theme="@style/AppTheme">
-        <activity android:name=".MainActivity"
+    <application
+        android:allowBackup="true"
```
This will not work since our software does not allow proprietary software (e.g. Google accounts) - https://developer.android.com/guide/topics/data/autobackup.html

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -0,0 +1,98 @@
+        manager = (WifiP2pManager) getSystemService(WIFI_P2P_SERVICE);
+        channel = manager.initialize(this, getMainLooper(), null);
+
+        new DiscoverAsyncTask().execute();
```
How can we trigger discovery without having to manually open the app?

sladomic replied:
I guess since we always have a master/slave connection we will need to try the discovery first (1-2 minutes) and if nothing is found then set this device as master and create a network. But I also found a "Mesh Network" idea here https://github.com/DrJukka/MyWifiMesh

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -0,0 +1,98 @@
+    private class DiscoverAsyncTask extends AsyncTask<Void, Void, Void> {
+
+        private boolean discover() {
+            Log.i(P.Tag, "discover called...");
```
Use `getClass().getName()` instead of `P.Tag`, so that we can see in the logcat which class the logging is coming from.

sladomic commented on app/src/main/java/org/literacyapp/synchronization/DeviceActionListener.java:
```diff
@@ -0,0 +1,15 @@
+public interface DeviceActionListener {
+    void showDetails(WifiP2pDevice device);
+    void cancelDisconnect();
+    void connect(WifiP2pConfig config);
```
maybe return a boolean when successful?

sladomic commented on app/src/main/java/org/literacyapp/synchronization/DeviceActionListener.java:
```diff
@@ -0,0 +1,15 @@
+    void showDetails(WifiP2pDevice device);
+    void cancelDisconnect();
+    void connect(WifiP2pConfig config);
+    void disconnect();
```
maybe return a boolean when successful?

sladomic commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -0,0 +1,240 @@
+        Log.d(P.Tag, "Peers Available");
+    }
+
+    private class DiscoverAsyncTask extends AsyncTask<Void, Void, Void> {
```
Move to another class (and maybe package, e.g. "task/DiscoverAsyncTask")

elich11 replied:
Yes, worth looking into it (Mesh Network).

sladomic commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -0,0 +1,240 @@
+            Log.i(P.Tag, "discover called...");
+            if (!isWifiP2pEnabled) {
+                Log.w(P.Tag, "isWifiP2pEnabled false");
+                return true;
```
What does it mean, if you return true here? Maybe add a short JavaDoc..

sladomic commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -0,0 +1,240 @@
+
+    }
+
+    public class WiFiDirectBroadcastReceiver extends BroadcastReceiver {
```
Move to another package/class (e.g. "receiver/WiFiDirectBroadcastReceiver")

sladomic commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -0,0 +1,240 @@
+                    // Wifi Direct mode is enabled
+                    activity.setIsWifiP2pEnabled(true);
+                    Log.d(P.Tag, "autoDiscover: " + autoDiscover);
+                    if (autoDiscover) {
```
Where is the autoDiscover variable set?

elich11 commented:
This is more a test then the actual production code (also a little test for me to see how to use the system).
The final code will work from a sticky service, will connect many2many, will be triggered from an alarm (all devices at the same time), will have a recovery mechanism.

jogrimst commented:
@elich11 Is this pull request ready to be merged?

elich11 committed:
peers connecting automatically

elich11 committed:
moved to sticky service

elich11 committed:
Added Status to UI

jogrimst commented:
@elich11 Do you need a code review and/or testing of the latest code changes?

elich11 commented:
For now, no need

________________________________
From: Jo Grimstad <notifications@github.com>
Sent: Thursday, February 2, 2017 8:54:44 PM
To: literacyapp-org/literacyapp-synchronization
Cc: elich11; Mention
Subject: Re: [literacyapp-org/literacyapp-synchronization] Moved to Service (sticky) added discovery timer (#6)


@elich11<https://github.com/elich11> Do you need a code review and/or testing of the latest code changes?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/literacyapp-org/literacyapp-synchronization/pull/6#issuecomment-277080509>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATjLRIuTZPCpO3EBRICcRl8YILVwgmuHks5rYkKUgaJpZM4Lj5RD>.

elich11 committed:
Added FileServerAsyncTask, FilesSendAsyncTast

elich11 committed:
Added mechanism to retry discovery after failed connection

elich11 committed:
Added force stop, start to options (Mostly for tests)

elich11 committed:
Added force stop, start to options (Mostly for tests)

elich11 committed:
push test file (start)

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -79,7 +81,23 @@ public void cancelDisconnect() {
 
     @Override
     public void connect(WifiP2pConfig config) {
+        Log.i(P.Tag, "Connect(WifiP2pConfig config) called.");
```
To reduce the amount of code, replace `P.Tag` with `getClass().getName()`

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main2Activity.java:
```diff
@@ -79,7 +81,23 @@ public void cancelDisconnect() {
     @Override
     public void connect(WifiP2pConfig config) {
+        Log.i(P.Tag, "Connect(WifiP2pConfig config) called.");
+        Toast.makeText(this,"Connecting..." ,Toast.LENGTH_LONG).show();
```
Remember to remove Toast messages before deploying the app to production. Consider logging it instead of using a Toast.

elich11 committed:
push test file (2)

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main3Activity.java:
```diff
@@ -3,14 +3,58 @@
+                        }
+                        try { Thread.sleep(5000); } catch (InterruptedException e) { e.printStackTrace(); }
+                        final String status = P.getStatus().toString();
+                        Log.d(P.Tag, "check state");
```
What is `P`?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/P.java:
```diff
@@ -5,5 +5,18 @@
+    public static final String Tag = "syncl";
+    private static Status mStatus = Status.Idle;
+
+    public enum Status {Idle,Discovering,FoundPeers, Connecting, Connected }
```
Make enums upper-case

jo-elimu reviewed (commented):
Can you add some JavaDoc explaining the responsibility of the WiFiDirectService?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/P.java:
```diff
@@ -1,11 +1,20 @@
  * Created by eli on 15/01/2017.
  */
 
 public class P {
```
What does `P` mean?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/P.java:
```diff
@@ -1,11 +1,20 @@
     public static final String Tag = "syncl";
+    public static final String TAG = "syncl";
+    private static String hostAdress = null;
+    public static final String DEFAULT_OUTPUT_FOLDER_NAME = "wifi_direct_files";
```
Rename to `.literacyapp-synchronization` to avoid images etc from appearing in Gallery app.

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/P.java:
```diff
@@ -18,5 +27,36 @@ public static Status getStatus() {
         return mStatus;
     }
 
+    public static void setOutputFolder(Context ctx, String outputFolder) {
```
Why is the output folder stored in SharedPreferences?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/WiFiDirectService.java:
```diff
@@ -294,4 +311,536 @@ public void onPeersAvailable(WifiP2pDeviceList wifiP2pDeviceList) {
+
+        public void close() {
+            isEnabled = false;
+            Log.i(P.TAG, "trying to close ServerSocket...");
```
Use `getClass().getName()` to make log debugging easier.

jogrimst commented:
@elich11 Which GitHub issue is this pull request for?: https://github.com/literacyapp-org/literacyapp-synchronization/projects

Can you make one pull request per GitHub issue? And please assign us for code review before the pull request contains too many changes. If you ask for code review after weeks of work and hundreds of code changes, it gets too hard to understand what the code is for and go through everything.

elich11 committed:
push test file, looks OK (first tests)

jo-elimu reviewed (commented):
@elich11 Are these changes for #9?

elich11 committed:
successful copy of test file.

jogrimst commented:
@elich11 I'm doing some testing of the app on a collection of tablets now. Is it necessary that all the tablets be connected to an existing Wi-Fi network?

elich11 commented:
Hi Jo,

Don't waste your time, its currently too unstable for testings

________________________________
From: Jo Grimstad <notifications@github.com>
Sent: Wednesday, February 22, 2017 7:16:17 AM
To: literacyapp-org/literacyapp-synchronization
Cc: elich11; Mention
Subject: Re: [literacyapp-org/literacyapp-synchronization] Added mechanism to retry discovery after connection fail (#6)


@elich11<https://github.com/elich11> I'm doing some testing of the app on a collection of tablets now. Is it necessary that all the tablets be connected to an existing Wi-Fi network?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/literacyapp-org/literacyapp-synchronization/pull/6#issuecomment-281588939>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATjLRCSuH8jfOPU8K0uu-OqyxP9oU6mEks5re-DBgaJpZM4Lj5RD>.

jogrimst commented:
@elich11 I'm asking because there will not be any Wi-Fi network available during the field testing in Tanzania. Only the tablets nearby each other. So the app will work without having to connect via an existing Wi-Fi router, right?

elich11 committed:
Added FileSent, FileReceived states, bug fixed, and improvements.

elich11 commented:
Yes.

________________________________
From: Jo Grimstad <notifications@github.com>
Sent: Wednesday, February 22, 2017 7:54:49 AM
To: literacyapp-org/literacyapp-synchronization
Cc: elich11; Mention
Subject: Re: [literacyapp-org/literacyapp-synchronization] Added mechanism to retry discovery after connection fail (#6)


@elich11<https://github.com/elich11> I'm asking because there will not be any Wi-Fi network available during the field testing in Tanzania. Only the tablets nearby each other. So the app will work without having to connect via an existing Wi-Fi router, right?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/literacyapp-org/literacyapp-synchronization/pull/6#issuecomment-281595288>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATjLRL2AiE0lPSuMh5sAnqkl8DFk8faTks5re-nJgaJpZM4Lj5RD>.

elich11 committed:
More fixes and stability improvements.
about added (with version).
Port moved to P

elich11 committed:
More fixes and stability improvements.
about added (with version).
Port moved to P

jo-elimu commented on app/build.gradle:
```diff
@@ -9,8 +9,8 @@ android {
-        versionCode 1000000
-        versionName "1.0.0"
+        versionCode 8000
+        versionName "0.8.0.00"
```
@elich11 Why do you use `0.8.0.00` as versionName? Please note that the versionCode has to be larger than the previous versionCode. If not, uploading the APK to the website will fail.

If you want to upgrade the version code after a small change, increase from 1.0.0 to 1.0.1.

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
> MAJOR version when you make incompatible API changes,
> MINOR version when you add functionality in a backwards-compatible manner, and
> PATCH version when you make backwards-compatible bug fixes.
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main3Activity.java:
```diff
@@ -75,10 +77,32 @@ public boolean onOptionsItemSelected(MenuItem item) {
+                        setMessage(getAboutText()).
+                        setPositiveButton(getString(R.string.ok), new DialogInterface.OnClickListener() {
+                            public void onClick(DialogInterface dialog, int id) {
+                                // User clicked OK button
```
Will the child have to press an OK button?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/Main3Activity.java:
```diff
@@ -75,10 +77,32 @@ public boolean onOptionsItemSelected(MenuItem item) {
         return super.onOptionsItemSelected(item);
     }
 
+    private String getAboutText() {
```
Is it necessary to include this?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/P.java:
```diff
@@ -24,6 +25,7 @@
     public static final String testFileName = "test.jpg";
     public static final String SENDER = "sender";
     public static final String RECEIVER = "receiver";
+    public static final int PORT = 8988;
```
Will all the tablets use the same port number?

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/WiFiDirectService.java:
```diff
@@ -438,7 +438,7 @@ private void sendTestFile(String hostAddress) {
         List<File> l = new ArrayList<File>();
         l.add(testFile);
-        new FilesSendAsyncTask(getApplicationContext(), l, hostAddress, 8988).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+        new FilesSendAsyncTask(getApplicationContext(), l, hostAddress, P.PORT).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
```
What is `P` in P.PORT?

elich11 committed:
new strategy to determine sender / receiver using groupOwner

elich11 committed:
new strategy to determine sender / receiver using groupOwner
can also force sender / receiver.

elich11 committed:
removed random generator from onStart
new strategy to determine sender / receiver using groupOwner
can also force sender / receiver.

jo-elimu commented on app/build.gradle:
```diff
@@ -9,8 +9,8 @@ android {
-        versionCode 8000
-        versionName "0.8.0.00"
+        versionCode 8100
+        versionName "0.8.1.00"
```
Use 3-digit versionName, like we do in all the other apps/games - http://semver.org
And use 7-digit versionCode: 1.2.3 = 1002003

![screen shot 2017-02-27 at 20 34 19](https://cloud.githubusercontent.com/assets/1451036/23372318/2746f33c-fd2c-11e6-8e60-3469f2168414.png)

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/WiFiDirectService.java:
```diff
@@ -383,12 +388,20 @@ public void onConnectionInfoAvailable(WifiP2pInfo info) {
-                Log.i(P.Tag, "senderReceiverType is reciever not sending test file");
+            }
+            else {
+                Log.w(P.TAG, "Group not formed yet...");
```
Use `getClass().getName()` instead of `P.TAG` to make it easier to read and debug log output.

jo-elimu reviewed (commented):
I've tested setting up the connection between 2 Pixel C tablets. An empty file `/sdcard/wifi_direct_files/wifip2pshared` was created on the receiving tablet.

jogrimst commented:
The application crashed on Pixel C:

```
02-28 14:13:05.142  2232  3302 D syncl   : DiscoverAsyncTask doInBackground
02-28 14:13:05.142  2232  3302 I syncl   : discover called...
02-28 14:13:05.144  2232  2232 I syncl   : discoverPeers onSuccess
02-28 14:13:06.531   512   575 D WifiStateMachine: starting scan for "JDIK"-WPA_PSK with 2437
02-28 14:13:06.765 10015 10015 D SystemWebChromeClient: file:///android_asset/www/90635dd228c9c19e3bd2.js: Line 26 : xtest: ftp: connect error=java.net.SocketTimeoutException: failed to connect to /192.168.0.1 (port 21) after 10000ms
02-28 14:13:06.765 10015 10015 I chromium: [INFO:CONSOLE(26)] "xtest: ftp: connect error=java.net.SocketTimeoutException: failed to connect to /192.168.0.1 (port 21) after 10000ms", source: file:///android_asset/www/90635dd228c9c19e3bd2.js (26)
02-28 14:13:07.027   512  1050 I ActivityManager: Process com.android.deskclock (pid 1327) has died
02-28 14:13:13.344   512  1008 D GraphicsStats: Buffer count: 6
02-28 14:13:13.344   512   655 I ActivityManager: Process org.literacyapp.synchronization (pid 2232) has died
02-28 14:13:13.344   512   655 W ActivityManager: Scheduling restart of crashed service org.literacyapp.synchronization/.WiFiDirectService in 1000ms
02-28 14:13:14.358   512   526 I ActivityManager: Start proc 1354:org.literacyapp.synchronization/u0a115 for service org.literacyapp.synchronization/.WiFiDirectService
02-28 14:13:14.372  1354  1354 I art     : Late-enabling -Xcheck:jni
02-28 14:13:14.434  1354  1354 W System  : ClassLoader referenced unknown path: /data/app/org.literacyapp.synchronization-2/lib/arm64
02-28 14:13:14.446  1354  1354 I syncl   : WiFiDirectService start sticky
02-28 14:13:14.446  1354  1354 D AndroidRuntime: Shutting down VM
02-28 14:13:14.448  1354  1354 E AndroidRuntime: FATAL EXCEPTION: main
02-28 14:13:14.448  1354  1354 E AndroidRuntime: Process: org.literacyapp.synchronization, PID: 1354
02-28 14:13:14.448  1354  1354 E AndroidRuntime: Theme: themes:{}
02-28 14:13:14.448  1354  1354 E AndroidRuntime: java.lang.RuntimeException: Unable to start service org.literacyapp.synchronization.WiFiDirectService@b903110 with null: java.lang.NullPointerException: Attempt to invoke virtual method 'android.os.Bundle android.content.Intent.getExtras()' on a null object reference
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3061)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.app.ActivityThread.-wrap17(ActivityThread.java)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1461)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:148)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:5461)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.os.Bundle android.content.Intent.getExtras()' on a null object reference
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at org.literacyapp.synchronization.WiFiDirectService.onStartCommand(WiFiDirectService.java:114)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3044)
02-28 14:13:14.448  1354  1354 E AndroidRuntime: 	... 8 more
02-28 14:13:14.510   512  1408 D OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: true
02-28 14:13:14.515   512   527 W HardwareRenderer: Could not acquire gfx stats buffer
02-28 14:13:14.515   512   527 W HardwareRenderer: android.os.RemoteException: Invalid package name
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at com.android.server.GraphicsStatsService.requestBufferForProcess(GraphicsStatsService.java:97)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.ThreadedRenderer$ProcessInitializer.initGraphicsStats(ThreadedRenderer.java:459)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.ThreadedRenderer$ProcessInitializer.init(ThreadedRenderer.java:447)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.ThreadedRenderer.<init>(ThreadedRenderer.java:126)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.HardwareRenderer.create(HardwareRenderer.java:436)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.ViewRootImpl.enableHardwareAcceleration(ViewRootImpl.java:759)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.ViewRootImpl.setView(ViewRootImpl.java:492)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:310)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:86)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.app.Dialog.show(Dialog.java:319)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at com.android.server.am.ActivityManagerService$UiHandler.handleMessage(ActivityManagerService.java:1489)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.os.Handler.dispatchMessage(Handler.java:102)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.os.Looper.loop(Looper.java:148)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at android.os.HandlerThread.run(HandlerThread.java:61)
02-28 14:13:14.515   512   527 W HardwareRenderer: 	at com.android.server.ServiceThread.run(ServiceThread.java:46)
02-28 14:13:14.549   512  1408 I OpenGLRenderer: Initialized EGL, version 1.4
02-28 14:13:17.221   512  9860 D NetlinkSocketObserver: NeighborEvent{elapsedMs=62368282, 192.168.0.1, [FC4AE9DEAD05], RTM_NEWNEIGH, NUD_STALE}
```

elich11 committed:
added test file

elich11 committed:
Some bug fixes related to sender / receiver

elich11 committed:
folder support (start)

elich11 committed:
folder support (start)

elich11 committed:
folder support working.

elich11 committed:
folder support working.
fixed state to reflect also folder support.

elich11 committed:
Added GroupDeleteHelper to delete persistent groups.
It seems that on some devices the old groups must be deleted, or the connection fails.

elich11 committed:
* Version updated.
* Added GroupDeleteHelper to delete persistent groups.
It seems that on some devices the old groups must be deleted, or the connection fails.

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/GroupDeleteHelper.java:
```diff
@@ -0,0 +1,95 @@
+import android.util.Log;
+import android.net.wifi.p2p.WifiP2pManager.Channel;
+
+public class GroupDeleteHelper   {
```
@elich11 What does "group" mean in this context? Can you add some JavaDoc explaining how groups work?

elich11 replied:
Done.

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/GroupDeleteHelper.java:
```diff
@@ -63,12 +64,16 @@ public Object invoke(Object proxy, java.lang.reflect.Method method, Object[] arg
-													Log.e(P.Tag, "Persistent Groups not deleted");
+													Log.e(P.Tag, "Persistent Group not deleted");
+													if (isUI)
+														Toast.makeText(ctx,"Persistent Group delete failed",Toast.LENGTH_LONG).show();
```
@elich11 Use Log output instead of Toast. We don't want Toasts to appear in the interface and confuse the children while they are busy learning something.

elich11 replied:
Thanks for your review.
Toasts will not appear when used by the children since they will be blocked by the isUI param.
I need the toasts internally, for debugging.

jo-elimu commented on app/src/main/java/org/literacyapp/synchronization/GroupDeleteHelper.java:
```diff
@@ -82,6 +87,7 @@ public void onFailure(int i) {
 					WifiP2pManager.class.getDeclaredMethod("requestPersistentGroupInfo", new Class[]{Channel.class, persistentInterface});
 
 			requestPersistentGroupMethod.invoke(mManager, mChannel, persitentInterfaceObject);
+			Toast.makeText(ctx,"Persistent Groups deleted",Toast.LENGTH_LONG).show();
```
Replace with logging

elich11 committed:
* Controller start

elich11 committed:
* Controller second update , but working

elich11 committed:
* Controller second update, not finished,  but working

elich11 committed:
* Flow diag added (controller)

jo-elimu reviewed (commented):
I uploaded this diagram to Google Drive, so that it can be opened and viewed in LucidChart: https://drive.google.com/drive/u/0/folders/0B4snxLj0L9wjZVBpbTlvMWhQQ3c

jogrimst commented:
@elich11 What will be the responsibility of the controller? Can you add some info about this in the JavaDoc of the class, or in a GitHub issue?

elich11 committed:
* Docs improved.
* Impl cont (2)
* Flow diag finished

elich11 committed:
* Impl cont (3)
* api target upgraded to 25 (build tools etc)

elich11 committed:
* Impl cont (4)

elich11 committed:
* Impl cont (5)

elich11 committed:
* Impl cont (6)

elich11 committed:
* Impl finished, passed sanity tests

elich11 committed:
* bug fixes (1)

elich11 committed:
* bug fixes (2)
* controller sanity2 passed.

elich11 committed:
* Version fixed.
* Added preferences screen.

elich11 committed:
* Alarm and configurations (2)

elich11 committed:
* Alarm and configurations (3)

elich11 committed:
* 1.0.3

elich11 committed:
Merge branch 'master' into master

elich11 commented:
upgraded to gradle 2.3

elich11 committed:
* 1.0.3

elich11 committed:
Merge remote-tracking branch 'origin/master'
